### PR TITLE
feature: add removeLayer function to View

### PR DESCRIPTION
**Pull request by Jeremy-Gaillard**
Added a function to remove a layer from a view.

How to test:
- Open examples/planar.html
- Run the command `itowns.View.prototype.removeLayer.call(view, 'planar')`

### CODE REVIEW

**Comment by Jeremy-Gaillard** [comment]
I'm having trouble re-instantiating the layer after removing it, so don't merge yet.

**Comment by Jeremy-Gaillard** [comment]
Never mind, the problem was elsewhere. The PR should work fine.

File: src/Core/View.js
Comment (autra) [inline src/Core/View.js:293]: If I'm not mistaken, obj are not necessarily added to `this.scene` any more (see GeometryLayer constructor).

**Review by autra — CHANGES_REQUESTED** [review]
I'm wondering if we can really do things that way for 2 reasons:
1. Layer can have parents. Btw have you looked at `layer.detach`? At the minimum `removeLayer` should properly detach from parent. What happens to children of the layer you're removing? They should get removed as well maybe?
1. Maybe we should call (or better: schedule a call for the next update loop) a `layer.cleanUp` method to give a chance to the layer to clean itself (remove cache, objects to the scene etc...). This call would recursively call cleanUp on children layers. The reason for this: intuitively I would think this cleaning process is particular to each protocol (to each *Processing actually).

I feel that things are not that simple here. On the other hand, what exact problem are you trying to solve right now? If you only need to disable/hide a layer, this could be very simple to implement (if not done already).

**Comment by Jeremy-Gaillard** [comment]
> Layer can have parents

The removeLayer function (in its current form) should only be used on layers that do not depend on a parent (i.e. only those in the _layers array). But you are right about the children needing to be properly detached.

> Maybe we should call a layer.cleanUp method

It would be a good idea indeed.

> I feel that things are not that simple here

I agree, part of the reason I created this PR was to gather feedback.

> what exact problem are you trying to solve right now?

I'm trying to give the user the possibility to reload a dataset with a different configuration. Technically, it means replacing a 3D Tiles layer with another 3D Tiles layer. Disabling or hiding the previous layer(s) doesn't seem very practical for this use case.

File: src/Core/View.js
Comment (Jeremy-Gaillard) [reply src/Core/View.js:293]: True. I should use `layer.object3d` instead of `this.scene`.

**Comment by autra** [comment]
Basically, if I understand correctly, removeLayer should undo what is relevant to undo in `addLayer`, `Layer.{attach,preUpdate,update,postUpdate}` and `Provider.preprocessLayer` (either in itself, or in layer.cleanup). That way, if we call again `addLayer` with the removed layer object, everything would work as expected. Maybe some info can stay (metadata gathered in preprocessLayer for instance, if we think they'll stay relevant), but on overall, itowns should go back to (nearly) the same state it has before the first `addLayer`. What do you think about these principles?

> I'm trying to give the user the possibility to reload a dataset with a different configuration

Interesting! Could you be a bit more specific about what you mean by "reloading a dataset with a different config"? Which configuration options are you talking about?

**Comment by Jeremy-Gaillard** [comment]
Your description seems correct. Though I don't think it's necessary to keep any metadata at all.

> Could you be a bit more specific about what you mean by "reloading a dataset with a different config"

Basically, it's changing the way the data is represented (lod, data type, ...). I'll share a demo of my work soon.

**Comment by gchoqueux** [comment]
Warning `GlobeView` have a method `removeLayer`, it's necessary to factorize the code

**Comment by peppsac** [comment]
@Jeremy-Gaillard maybe we could close this PR in favor of #546 ?

**Comment by Jeremy-Gaillard** [comment]
Indeed we can. Closing.

**Pull request closed without merging**
